ActiveSupport::Cache::Key object - Faster Cache Keys

schneems opened the pull request:
This PR moves the existing logic of `expanded_key` and `expanded_version` methods from ActiveSupport::Cache to their own class. This PR is a performance refactor.

Currently, when a cache entry is fetched the method, `normalize_key` (which calls `ecpanded_key`) is called on the key multiple times both for generating the actual key and for use with the cache store and logging. The existing version of this method is quite expensive and allocates memory. The biggest example is that it generates new arrays when the input is an array by calling `compact` and then again by calling `Array#to_param`. By moving the required logic to generate a key to a class, the need to generate intermediate array objects is drastically reduced. The key is generated as elements are added to the object as opposed to when they are needed. This behavior alleviates the need to re-generate the same key value for cache retrieval and logging.

One crucial property of this new cache class is that keys are composable. This composability means that if you pass the same objects into a class instance in the same order, it will produce the same key. This property allows us to optimize cache key generation in `collection_caching.rb` as the root of all elements being cached are the same.

## Performance

### Memory

Before:

```
Total allocated: 741750 bytes (6645 objects)
Total retained:  18917 bytes (166 objects)
```

After:

```
Total allocated: 707548 bytes (6006 objects)
Total retained:  18917 bytes (166 objects)
```

Diff:

```
(741750 - 707548) / 741750 # => ~4.6 %
```

### Performance

Here's the raw results: https://gist.github.com/schneems/7e808e7a866c1bc3a2d50b5e0ae4db88 for 63 runs, each with 1000 hits to the index of CodeTriage.

On average this patch is **7.24%** faster than the commit before it. This number is statistically significant.

schneems committed:
ActiveSupport::Cache::Key object - Faster Cache Keys

This PR moves the existing logic of `expanded_key` and `expanded_version` methods from ActiveSupport::Cache to their own class. This PR is a performance refactor.

Currently, when a cache entry is fetched the method, `normalize_key` (which calls `ecpanded_key`) is called on the key multiple times both for generating the actual key and for use with the cache store and logging. The existing version of this method is quite expensive and allocates memory. The biggest example is that it generates new arrays when the input is an array by calling `compact` and then again by calling `Array#to_param`. By moving the required logic to generate a key to a class, the need to generate intermediate array objects is drastically reduced. The key is generated as elements are added to the object as opposed to when they are needed. This behavior alleviates the need to re-generate the same key value for cache retrieval and logging.

One crucial property of this new cache class is that keys are composable. This composability means that if you pass the same objects into a class instance in the same order, it will produce the same key. This property allows us to optimize cache key generation in `collection_caching.rb` as the root of all elements being cached are the same.

## Performance

### Memory

Before:

```
Total allocated: 741750 bytes (6645 objects)
Total retained:  18917 bytes (166 objects)
```

After:

```
Total allocated: 707548 bytes (6006 objects)
Total retained:  18917 bytes (166 objects)
```

Diff:

```
(741750 - 707548) / 741750 # => ~4.6 %
```

### Performance

Here's the raw results: https://gist.github.com/schneems/7e808e7a866c1bc3a2d50b5e0ae4db88 for 63 runs, each with 1000 hits to the index of CodeTriage.

On average this patch is **7.24%** faster than the commit before it. This number is statistically significant.

schneems commented:
Apparently all the tests are currently passing, that's a bit of a surprise.

schneems commented on actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb:
```diff
@@ -54,13 +54,21 @@ def callable_cache_key?
       def collection_by_cache_keys
         seed = callable_cache_key? ? @options[:cached] : ->(i) { i }
 
+        key_base = ActiveSupport::Cache::KeyMaster.new(@view.combined_fragment_cache_key(nil)).cache_key
```
One thing I don't know here is if the `@view.combined_fragment_cache_key(nil)` can return back any objects that respond to `:cache_version`. If so we need to also output make the keymaster object able to initialize with a default cache key as well. Should be an easy-ish enough.

rafaelfranca commented on activesupport/lib/active_support/cache.rb:
```diff
@@ -18,6 +18,65 @@ module Cache
+        end
+      end
+
+      private def store_key(key)
```
please don't use this style in the rails codebase. 🙏

schneems replied:
Thank you for the review, I will change it.

I will also tell you why I prefer this style. 

Since many rails classes are very long, it is hard to know what method is public and private while you are reading the code, especially if it has no documentation and isn't explicitly `:nodoc:`, which most code isn't.

It looks like it is possible to add a rule to rubocop to detect this issue https://rubocop.readthedocs.io/en/latest/cops_style/#styleaccessmodifierdeclarations. Should it be added to the rules?

rafaelfranca replied:
Good idea. :+1: for that rule.

rafaelfranca commented on activesupport/lib/active_support/cache.rb:
```diff
@@ -18,6 +18,65 @@ module Cache
+          key.sort_by! { |k, _| k.to_s }
+          key.collect! { |k, v| "#{k}=#{v}" }
+          key.each do |k|
+            append_part(key)
```
Where is `append_part` defined?

rafaelfranca commented on activesupport/lib/active_support/cache.rb:
```diff
@@ -18,6 +18,65 @@ module Cache
+      attr_reader :cache_key, :cache_version
+
+      def initialize(key = nil)
+        @cache_key = String.new
```
Do you want to create a string with ASCII encoding or UTF8 encoding?

schneems replied:
Unsure, we probably want to handle UTF 8 cache keys.

rafaelfranca replied:
Yeah, so I think we should be using `+""`. String.new changes the encoding of the string.

```
>> String.new.encoding
=> #<Encoding:ASCII-8BIT>
>> "".encoding
=> #<Encoding:UTF-8>
```

rafaelfranca reviewed (commented):
The implementation looks way more complicated right now. It is really hard to know what this new object does and why it exists but I guess in the final patch, given this object will replace existing methods it will be easier to know the advantages to the code.

oniofchaos commented on actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb:
```diff
@@ -54,13 +54,21 @@ def callable_cache_key?
+        key_base = ActiveSupport::Cache::KeyMaster.new(@view.combined_fragment_cache_key(nil)).cache_key
         @collection.each_with_object({}) do |item, hash|
-          hash[expanded_cache_key(seed.call(item))] = item
+          key = ActiveSupport::Cache::KeyMaster.new(key_base)
```
If I'm reading this right, it seems like you're re-creating an instance of `ActiveSupport::Cache::KeyMaster.new(key_base)` for each object of the loop, when `key_base` is a static value that's declared outside the loop.

Could you also declare `key` outside the loop and then keep the `key << ...` line inside the loop?

schneems replied:
No because then we would be using the same object for every entry in the hash and then mutating it. If you do that you get AMAZING performance gains, but that's mostly because you end up with only one item in the hash since all of them are using the same key.

schneems commented:
@rafaelfranca @oniofchaos @georgeclaghorn i'm mostly done experimenting here. Ready for the next round of reviews.

jeremy commented on activesupport/lib/active_support/cache.rb:
```diff
@@ -449,7 +453,7 @@ def fetch_multi(*names)
       def write(name, value, options = nil)
         options = merged_options(options)
-
+        name = KeyMaster.to_key(name)
```
Intentional newline was removed. Please restore it.

jeremy commented on activesupport/lib/active_support/cache/key_master.rb:
```diff
@@ -0,0 +1,165 @@
+
+module ActiveSupport
+  module Cache
+    # This class is responsible for cohercing input into a valid
```
`coercing`

jeremy commented on activesupport/lib/active_support/cache/key_master.rb:
```diff
@@ -0,0 +1,165 @@
+    # When generating a a cache key, `nil` will be filtered
+    # out, so `["ghost", nil "busters"]` will generate the
+    # same output as ["ghost", "busters"]. If presence of a
+    # `nil` is important, then convert it to a string.
```
Is this a behavior change? `nil` is typically meaningful. Removing it causes collisions.

jeremy commented on activesupport/lib/active_support/cache/key_master.rb:
```diff
@@ -0,0 +1,165 @@
+        return key if key.is_a?(KeyMaster)
+
+        self.new(key)
+      end
```
Would compare before/after usage with `.for(key)` or simply `.new(key)`. The `to_*` methods indicate that the receiver is being coerced, not the argument.

No need for a guard here, since it's expected that a key may be an instance of this class. A conditional expression is appropriate.

jeremy commented on activesupport/lib/active_support/cache/key_master.rb:
```diff
@@ -0,0 +1,165 @@
+        self << key
+      end
+
+      def <<(key)
```
Suggest using Digest-like `#update` (with a `#<<` alias) for clearer internal usage.

jeremy commented on activesupport/lib/active_support/cache/key_master.rb:
```diff
@@ -0,0 +1,165 @@
+      end
+
+      def <<(key)
+        return if key.nil?
```
`nil` keys are significant.

schneems replied:
Yeah? I think they're currently filtered out via `compact` in the current implementation of `Array#to_param`. I can add a test to confirm.

schneems replied:
Great catch. We should definitely have this tested then https://github.com/rails/rails/pull/34128

jeremy commented on activesupport/lib/active_support/cache/key_master.rb:
```diff
@@ -0,0 +1,165 @@
+
+      private
+        def store_key(key)
+          @cache_key << "/" unless @cache_key.empty?
```
Empty string keys are significant, so checking for emptiness will be insufficient.

schneems replied:
This line is doing the work of `Array#to_param` which does a `join("/")` on the elements. We want a `/` separator between each entry, but we don't want to start the key with a slash if it's the first time an entry is added.

jeremy replied:
Understood. The implementation is not equivalent, though. Consider adding a key that is an empty string: `@cache_key` will still be empty and the next key that's added will omit the `/` that should be present.

jeremy commented on activesupport/test/cache/behaviors/cache_store_behavior.rb:
```diff
@@ -21,8 +21,9 @@ def test_fetch_without_cache_miss
-      assert_equal "baz", @cache.fetch("foo") { "baz" }
+    key = ActiveSupport::Cache::KeyMaster.new("foo")
+    assert_called_with(@cache, :write, [key, "baz", @cache.options]) do
+      assert_equal "baz", @cache.fetch(key) { "baz" }
```
Is this meant to incidentally test with the new key instances? Or to demonstrate their usage?

Think we'll likely want to leave these tests demonstrating typical usage with string keys and introduce tests asserting that the cache stores behave as expected with new key instances.

schneems replied:
The reason I modified this is the `assert_called_with` uses the exact value and in this case `write` receives an instance of `KeyMaster`. Previously it received an instance of a string. I couldn't find another way to re-write this without directly passing in the object. Otherwise, if we pass in a regular string it gets converted to the key object so the mock fails. If you've got ideas here i'm open to suggestion.

jeremy replied:
Got it! 👍

jeremy commented on activesupport/test/cache/key_master_test.rb:
```diff
@@ -0,0 +1,60 @@
+require "abstract_unit"
+require "active_support/cache"
+
+class KeyMasterTest < ActiveSupport::TestCase
```
Looks like we'll need test coverage for nil and empty keys.

jeremy reviewed (changes requested):
Solid concept and approach. Thanks for digging deep on this @schneems.

schneems commented on actionview/lib/action_view/renderer/partial_renderer/collection_caching.rb:
```diff
@@ -54,13 +54,16 @@ def callable_cache_key?
-          hash[expanded_cache_key(seed.call(item))] = item
+          key = ActiveSupport::Cache::KeyMaster.new(key_base)
+          key << cache_fragment_name(seed.call(item))
+          hash[key] = item
```
I would still probably want it to be the same number of lines. Something like 

```
name = cache_fragment_name(seed.call(item))
key = ActiveSupport::Cache::KeyMaster.new(key_base, name)
hash[key] = item
```

Otherwise it gets a bit hard to read

```
key = ActiveSupport::Cache::KeyMaster.new(key_base, cache_fragment_name(seed.call(item)))
hash[key] = item
```

If you don't mind option 2 let me know. I'll need to modify the method signature to take a secondary argument. I don't want to use `*args` as I don't want to allocate an extra array per invocation.

jeremy replied:
Not golfing for length of method here, but for coherence and readability. This method isn't concerned with how to construct a key; `expanded_cache_key` is.

tenderlove commented on activesupport/lib/active_support/cache.rb:
```diff
@@ -311,29 +312,31 @@ def mute
+            get_entry_value(entry, master_key, options)
           else
-            save_block_result_to_cache(name, options) { |_name| yield _name }
+            save_block_result_to_cache(master_key, options) { yield raw_key }
```
Looks like the block param was introduced here: https://github.com/rails/rails/commit/7f6017c4a8f

Doesn't look like there was any reason it was introduced.

jeremy replied:
That moved the code. Origin looks like #8013

tenderlove commented:
How do I reproduce the numbers?  I tried `rake -f perf.rake perf:test` in codetriage, but I get an exception.

schneems commented on activesupport/lib/active_support/cache.rb:
```diff
@@ -311,29 +312,31 @@ def mute
       #   cache.fetch('foo') # => "bar"
-      def fetch(name, options = nil)
+      def fetch(raw_key, options = nil)
+        master_key = KeyMaster.to_key(raw_key)
```
In most methods you can safely replace the incoming name variable with an instance of KeyMaster, however here we are using the raw key by passing it into the block so it cannot change. 

After that I wanted to preserve passing in a KeyMaster object to all places that would otherwise have taken a `name`.

Naming is hard. I'm open to suggestions.

oniofchaos commented on activesupport/lib/active_support/cache/key.rb:
```diff
@@ -0,0 +1,160 @@
+      attr_reader :cache_key, :cache_version
+
+      def self.to_key(key)
+        return key if key.is_a?(Key)
```
Do you happen to have a feel for which scenario (passing in a `Key` vs something else) is more common? I did a microbenchmark (admittedly with strings/arrays) and a ternary would be faster for the else condition.

```
one - the way you have now
two - ternary
Warming up --------------------------------------
            one - if   184.196k i/100ms
            two - if   157.414k i/100ms
          one - else   172.947k i/100ms
          two - else   225.391k i/100ms
Calculating -------------------------------------
            one - if      5.276M (± 3.9%) i/s -     26.340M in   5.000433s
            two - if      3.418M (± 5.6%) i/s -     17.158M in   5.037682s
          one - else      3.710M (± 9.9%) i/s -     18.332M in   5.004147s
          two - else      6.629M (± 5.8%) i/s -     33.132M in   5.018285s

Comparison:
          two - else:  6629185.6 i/s
            one - if:  5275983.5 i/s - 1.26x  slower
          one - else:  3709665.6 i/s - 1.79x  slower
            two - if:  3417664.4 i/s - 1.94x  slower
```

schneems replied:
I'm not able to reproduce that. Can you please share your bench script when posting results. Here's mine https://gist.github.com/schneems/8412f1c85672c76fc9946111138d9e66.

> Do you happen to have a feel for which scenario (passing in a Key vs something else) is more common?

Ideally, we want to introduce a key object as soon as possible otherwise if a non `Key` object such as an array makes it all the way down to `normalize_key` then duplicate work will end up happening. That's why there are several cases where we're ensuring the key we are working with is a `Key` object.

oniofchaos replied:
https://gist.github.com/oniofchaos/297eaaad5ec3d0b4dde394797bcb34ed

oniofchaos replied:
Whoops....at any rate, looks like this code got refactored. Thanks.

oniofchaos commented on activesupport/lib/active_support/cache/key.rb:
```diff
@@ -0,0 +1,160 @@
+        if key.respond_to?(:cache_key)
+          store_key(key.cache_key.to_s)
+          store_version(key)
+          return
```
Could you do an `else` instead of having a `return` statement here?

oniofchaos commented on activesupport/lib/active_support/cache/key.rb:
```diff
@@ -0,0 +1,160 @@
+
+        case key
+        when Array
+          key.each do |k|
```
Seems like this could be one-lined?

schneems replied:
Agree, I like that better, thanks.

oniofchaos commented on activesupport/lib/active_support/cache/key.rb:
```diff
@@ -0,0 +1,160 @@
+          key.sort_by! { |k, _| k.to_s }
+          key.collect! { |k, v| "#{k}=#{v}" }
+
+          key.each do |k|
```
Seems like this loop plus the last loop could be combined.

```ruby
key.each do |k|
  store_key("#{k}=#{v}")
end
```

schneems replied:
I like that better, makes it cleaner, can do it on one line as well. Thanks!

oniofchaos commented on activesupport/lib/active_support/cache/key.rb:
```diff
@@ -0,0 +1,160 @@
+      private
+        def store_key(key)
+          @cache_key << "/" unless @empty_key
+          @empty_version =
```
This line seems weird, since there's nothing on the other side?

schneems replied:
Thanks, fixing

oniofchaos commented on activesupport/lib/active_support/cache/key.rb:
```diff
@@ -0,0 +1,160 @@
+          @empty_key = false
+        end
+
+        def store_version(key)
```
This method feels like it's handling a lot of different things at once - dealing with saving cache versions from individual keys as well as looping through array keys in order to recursively handle the cache version of each key. I wonder if it makes sense to handle those two responsibilities separately?

schneems replied:
I split `store_key` out to it's own method as it was called multiple places from `update`. I like that this logic for actually adding the key is inside the same method while all other conditions in the case are recursive, it helps me to distinguish between the two.

oniofchaos replied:
Hmm, lemme illustrate what I'm imagining in my head and if you like it better the way it is now then it's fine - more of a SRP issue than anything technical.

```ruby
def store_version(key)
  split_object(key) unless key.respond_to(:cache_version)
  @cache_version << "/" unless @cache_version.empty?
  @cache_version << key.cache_version.to_param.to_s
end

def split_object(key)
  if key.is_a?(Array)
    key.each { |k| store_version(k) }
  elsif key.respond_to?(:to_a) 
    split_object(key.to_a)
  end
end
```

oniofchaos commented on activesupport/lib/active_support/cache/key.rb:
```diff
@@ -0,0 +1,160 @@
+              store_version(k)
+            end
+          when key.respond_to?(:to_a)
+            store_version(key.to_a)
```
On my initial read at least, I felt like it would make sense to combine this with the above logic or duplicate the above logic so that you're not adding an extra layer of recursive logic.

schneems replied:
I don't totally follow here. This method represents the existing logic of `expanded_version` while the main body of `<<` represents the existing logic of `expanded_key`. 

We cannot get rid of this `to_a` behavior for the case of a hash containing another object that can respond to `to_a` such as another hash.

I'll add this as a test case to ensure this logic is preserved

```
    klass = Class.new { def cache_version; "foo"; end }
    key = ActiveSupport::Cache::Key.new({hello: {world: klass.new}})
    assert_equal "foo", key.cache_version
```

dillonwelch reviewed (commented):
Mostly on the nitpick/spelling side of things...

georgeclaghorn commented on activesupport/lib/active_support/cache/key.rb:
```diff
@@ -0,0 +1,160 @@
+    #   k1 << "stay-puft"
+    #   k2 = Key.new(["foo", "bar"])
+    #   k1.cache_key == k2.cache_key
+    #   # => true
```
Why would `k1.cache_key == k2.cache_key` be true? Should line 39 be `k1 << "bar"`?

schneems commented on activesupport/lib/active_support/cache/key.rb:
```diff
@@ -0,0 +1,160 @@
+    #   Key.new(klass.new).cache_key
+    #   # => "foo"
+    #
+    # In addition to generating a `cache_key`
```
👍

schneems commented on activesupport/lib/active_support/cache/key.rb:
```diff
@@ -0,0 +1,160 @@
+    #   k1.cache_key == k2.cache_key
+    #   # => true
+    #
+    # In addition to passing a key into the initialize method
```
👍

schneems commented on activesupport/lib/active_support/cache/key.rb:
```diff
@@ -0,0 +1,160 @@
+
+module ActiveSupport
+  module Cache
+    # This class is responsible for cohercing input into a valid
```
👍

schneems commented on activesupport/lib/active_support/cache/key.rb:
```diff
@@ -0,0 +1,160 @@
+    #
+    # It is faster to directly add objects to the
+    # Key object via `<<` than to generate
+    # an intermediate array of objects
```
This object is intended to be internal only, a warning when the user cannot do anything about it directly isn't great.

schneems commented:
Rebased all the commits. I left the cherry-pick from https://github.com/rails/rails/pull/34128 in in case we apply that separately.

jeremy commented on activesupport/lib/active_support/cache.rb:
```diff
@@ -312,28 +313,31 @@ def mute
       #   end
       #   cache.fetch('foo') # => "bar"
       def fetch(name, options = nil)
+        master_key = Key.call(name)
```
We have a few forms of keys floating around this method and the naming is piling up. How would you communicate the role of the "master" key here? How does it differ from the "normalized" key, or the name?

Seems we have key components (name), normalized key (serialized string key), and a Key instance (represents key components and serializes to a string).

In that case, 
- are we changing instrumentation behavior by passing the key instance rather than its components?
- should the Key instance be responsible for namespacing and "normalization" as well? Where's the boundary with cache store implementation concerns, like transforming the serialized keys?

schneems replied:
I'm still working my way through your comments. I wanted to go ahead and mention that I looked into this for a few hours:

> should the Key instance be responsible for namespacing and "normalization" as well? Where's the boundary with cache store implementation concerns, like transforming the serialized keys?

I was able to push the responsibility to the Key class, but we're not able to eliminate calls to the `normalize_key` method calls, since other cache stores such as the redis store, extend this method https://github.com/rails/rails/blob/5431e17733366da1fd10f2cd3039d66a56012683/activesupport/lib/active_support/cache/redis_cache_store.rb#L434-L436

Since we can't get rid of that method call, the read/write/fetch method calls end up looking similar, however the Key class becomes more convoluted. I think it was worthwile to look into and I had wondered if it was possible before. Right now it doesn't look like the benefits outweight the costs.

> are we changing instrumentation behavior by passing the key instance rather than its components?

We are not changing the instrumetation behavior. In the `instrument` method, it calls `normalize_key` on the argument that is passed. If you don't pass in a key object, one gets created. The resulting generated strings will be identical.

> We have a few forms of keys floating around this method and the naming is piling up. How would you communicate the role of the "master" key here? How does it differ from the "normalized" key, or the name?

I can do a better job of naming inside of the `fetch` method. This variable makes less sense now that the class is no longer "key master". 

I'm not ready for another review yet, but wanted to post my progress while it's fresh.

jeremy commented on activesupport/lib/active_support/cache/key.rb:
```diff
@@ -0,0 +1,161 @@
+        else
+          self.new(key)
+        end
+      end
```
Wondering whether overriding `Key.new` is warranted. Calling `Key.call` suggests I'm somehow invoking Key. What other precedent do we have in Ruby for selectively coercing objects of many kinds to an object of the desired kind? Toplevel methods like `Kernel#Array` and `#Integer` come to mind, and `Hash.[]`.

schneems replied:
Here's some other options for the `Key.call` method.

```ruby
key = Key.build(name)

key = Key(name)

key = Key[name]
```

I like the `Array()` style call `Key()` the best, since it's behavior that could return a new value or the same value. I wish there was a better english version, but every long attempt I try ends up too unwieldy.

If we stop trying to do the checks and instead simply call `Key.new` each time then we end up losing a signifigant amount of memory savings.

Alternatively we could get rid of this fancy method all together if we instead did this check at each call site:

```ruby
if name.is_a?(Key)
  key = name
else
  key = Key.new(name) 
end
```

That's a little too verbose for my tastes.

Alternatively we could push this behavior to a method inside of `cache.rb` but then we still have a naming issue. Could call it something like `ensure_key()`, `key_or_generate_key()`, `key_or_new()`, etc.

schneems replied:
I changed it to `Key()` however it's not a kernel method, so the full invocation has to be used `ActiveSupport::Cache::Key("foo")`. What do you think?

jeremy commented on activesupport/lib/active_support/cache/key.rb:
```diff
@@ -0,0 +1,161 @@
+    class Key # :nodoc:
+      attr_reader :cache_key, :cache_version
+
+      def self.call(key)
```
It sounds unwieldy, but an argument like `key_or_components` or `keyable` suggests that we expect and accept a wide range of arguments that quack like Key instances, or objects with cache_keys, or components of keys.

schneems replied:
I don't think that's unweildy at all, `key_or_components` looks good, I'll use that.

schneems replied:
I switched to the updated variable name.

jeremy commented on activesupport/lib/active_support/cache/key.rb:
```diff
@@ -0,0 +1,161 @@
+            split_version(key.to_a)
+          end
+        end
+    end
```
Implementation is feeling understandable. Still a challenge to grok that it's recursively, incrementally building a serialized key name and version.

Makes me dream of "just" using Digest classes for both! Breaking all compat for sure—but already built for efficient incremental digest generation.

schneems replied:
Agreed, there are lots of little edge cases here-and-there, where different keys can produce the same key. Overall it seems that's not caused major issues so it seems fine. To make issues worse, the `expand_cache_key` method has a slightly different implementation of how cache key expansion is done.

To add support for the latest dalli release I have to add `cache_key_with_version`. The implementation (not currently pushed) is not fully backwards compatable. For example if there are multiple keys instead of `<cache-key-1>-<cache-version-1>/<cache-key-2>-<cache-version-2>` we end up with `<cache-key-2>/<cache-key-2>-<cache-version-1>/<cache-version-2>`. We can make it backwards compat, but it will require storing and writing a third string.

jeremy commented on activesupport/test/cache/behaviors/cache_store_behavior.rb:
```diff
@@ -498,7 +500,41 @@ def test_cache_miss_instrumentation
+    def does_not_support_blank_keys?
+      @cache.is_a?(ActiveSupport::Cache::FileStore) ||
+        @cache.is_a?(ActiveSupport::Cache::MemCacheStore)
+    end
```
Let's leave these for the other PR and rely on Key unit tests, otherwise.

schneems replied:
Sounds good. I plan on deleting this commit and porting these to unit tests. Thanks for the review!

schneems replied:
I already had all the unit tests proposed in the other PR. I deleted some of the tests that produced a `nil` or `empty` key.

schneems replied:
I already had all the unit tests proposed in the other PR. I deleted some of the tests that produced a `nil` or `empty` key.

schneems replied:
I already had all the unit tests proposed in the other PR. I deleted some of the tests that produced a `nil` or `empty` key.

jeremy reviewed (commented):
Looking good! Loving that the resulting code is getting *easier* to understand than before.

schneems committed:
Generate expand_cache_key via ActiveSupport::Cache::Key

The current format of the cache key generated via `retrieve_cache_key` which is used by `expand_cache_key` is different than the other method of "expanding" a cache key provided in `expanded_key`. This method is expected to provide a raw cache key that changes as the version changes.

While this method currently relies on the specialized method `cache_key_with_version` this method is not documented in any of our guides and it is confusing that using a custom object with both a `cache_key` and a `cache_version` will not generate an appropriate cache key. 

This method `cache_key_with_version` is also used by the dalli store which resulted in some confusion https://github.com/petergoldstein/dalli/issues/710.

This commit moves all Rails internal cache key generation into the ActiveSupport::Cache::Key object so that if the same objects are used as a key and passed to `expand_cache_key` then it will match an identical key and object being pulled via `expanded_key` if the cache store is `:dalli_store`.

The downsides to this approach:

- The method `cache_key_with_version` now does nothing. Implementing it on an AR object or another custom object will not do anything.
- Anyone relying on this method before will now get cache keys invalidated as the format has changed (instead of `key1-value1/key2-value2` it will now be `key1/key2-value1/value2`).

The upside to this approach:

- Key generation is now completely centralized
- People writing custom objects do not need to know about a third method to implement in addition to `cache_key` and `cache_version`.

schneems commented:
@jeremy I have a proposed commit that will make this class compatible with `cache_key_with_version` and it also gets rid of custom code in `expand_cache_key`.

The major downside is this would change the format of keys that are currently going through that path. This commit makes this proposal more invasive, but ultimately more compatible with things like the Dalli cache store. Here's the commit https://github.com/rails/rails/pull/33944/commits/62ee4a5473ddbc8bfc7d8edf632029a26ed5e620

schneems commented:
For `cache_key` support, I could instead retain a reference to the original objects which would be a backward compatible change. I'm not sure about the performance impact, likely it would still be faster. Would take some time to stub out and try out.

schneems commented:
Closing in favor of #36310
